### PR TITLE
[FSDP] Reduce CPU overhead

### PR DESCRIPTION
**Pull request by awgu**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #96958
* #96326

I experimented with 200 `nn.Linear`s with `bias=True` for a total of 400 `nn.Parameter`s all wrapped into the same FSDP instance and world size of 2.

**`unshard()` -> `_use_unsharded_views()`**
- (From previous PR) unsafe `setattr`: 6.112 ms -> 4.268 ms

**`pre_unshard()` -> `_writeback_orig_params()`**
- Factor out `flat_param` and `flat_param_grad` data pointers: ~1.8 ms -> 1.071 ms
    - Now dominated by calling `_typed_storage()` on each original parameter and its gradient

**`reshard()` -> `_use_sharded_views()`**
- Factor out `torch.empty(0, ...)`: ~4.6 - 4.7 ms -> ~2.7 - 2.8 ms
    - Now dominated by `aten::slice()` and (unsafe) `setattr`, which are required

I removed some `assert` calls that were only needed for mypy or if the subsequent call would provide the same error message anyway. These have negligible overhead, but I think it is still okay to remove them and avoid the type check. We need to address type checking more holistically anyway.

---

### CODE REVIEW

File: torch/distributed/fsdp/flat_param.py [hunk 1]
```diff
@@ -1573,31 +1561,30 @@ def _use_sharded_views(self) -> None:
...
                 offset += numel_in_shard
             else:
                 # Allow the original data to be freed via garbage collection
-                param.data = torch.empty(
-                    0,
-                    dtype=self.flat_param.dtype,  # in case `flat_param` changed dtype
-                    device=self.flat_param.device,
-                    requires_grad=False,
-                )
+                param.data = size_0_empty_tensor
```
Comment (rohan-varma) [inline torch/distributed/fsdp/flat_param.py:1587]: now each param's data will point to the same tensor's storage, I guess that's okay because this tensor doesn't really matter and is just used to allow the original data to be freed?

**Review by rohan-varma — APPROVED** [review]
Awesome investigation, LGTM!

**Comment by awgu** [comment]
All failures unrelated:

manywheel-py3_8-cuda11_7-test / test
```
2023-03-16T18:06:44.0970952Z ␛[?25hERROR: Could not find a version that satisfies the requirement pytorch-triton==2.1.0+b8b470bc59 (from torch) (from versions: 0.0.1, 2.0.0+0d7e753227, 2.0.0+3aa3d7024e, 2.0.0+af76c989eb, 2.0.0+b8b470bc59, 2.0.0+c8bfe3f548, 2.0.0+d54c04abe2, 2.1.0+2c32f43999)
2023-03-16T18:06:44.0973681Z ERROR: No matching distribution found for pytorch-triton==2.1.0+b8b470bc59
```

win-vs2019-cpu-py3 / test (default, 2, 2, windows.4xlarge)
```
FAILED test_ops_jit.py::TestJitCPU::test_variant_consistency_jit_diagonal_scatter_cpu_float32
```

win-vs2019-cuda11.7-py3 / test (default, 3, 5, windows.g5.4xlarge.nvidia.gpu)
```
FAILED test_ops_jit.py::TestJitCUDA::test_variant_consistency_jit_nanquantile_cuda_float32
```
